README.md
convertdate
===========

The convertdate package was originally developed as "[Python Date
Utils](http://sourceforge.net/projects/pythondateutil/)" by Phil
Schwartz. It had been significantly updated and expanded.

[Consult the complete docs for detailed usage info](https://convertdate.readthedocs.io/en/latest/).

Available calendars:

-   Armenian
-   Bahai
-   Coptic (Alexandrian)
-   French Republican
-   Gregorian
-   Hebrew
-   Indian Civil
-   Islamic
-   Julian
-   Mayan
-   Persian
-   Positivist
-   Mayan
-   ISO
-   Ordinal (day of year)
-   Dublin day count
-   Julian day count

The `holidays` module also provides some useful holiday-calculation,
with a focus on North American and Jewish holidays.

Installing
----------

`pip install convertdate`

Or download the package and run `python setup.py install`.

Using
-----

    >>> from convertdate import french_republican
    >>> from convertdate import hebrew
    >>> french_republican.from_gregorian(2014, 10, 31)
    (223, 2, 9)
    >>> hebrew.from_gregorian(2014, 10, 31)
    (5775, 8, 7)

Note that in some calendar systems, the day begins at sundown.
Convertdate gives the conversion for noon of the day in question.

Each module includes a `monthcalendar` function, which will generate a
calender-like nested list for a year and month (each list of dates runs
from Sunday to Saturday)

    >>> hebrew.monthcalendar(5775, 8)
    [
        [None, None, None, None, None, None, 1],
        [2, 3, 4, 5, 6, 7, 8],
        [9, 10, 11, 12, 13, 14, 15],
        [16, 17, 18, 19, 20, 21, 22],
        [23, 24, 25, 26, 27, 28, 29]
    ]

    >>> julian.monthcalendar(2015, 1)
    [
       [None, None, None, 1, 2, 3, 4],
       [5, 6, 7, 8, 9, 10, 11],
       [12, 13, 14, 15, 16, 17, 18],
       [19, 20, 21, 22, 23, 24, 25],
       [26, 27, 28, 29, 30, 31, None]
    ]

Special Options
---------------

### Armenian

The Armenian calendar begins on 11 July 552 (Julian) and has two modes of
reckoning. The first is the invariant-length version consisting of 12 months
of 30 days each and five epagomenal days; the second is the version
established by Yovhannes Sarkawag in 1084, which fixed the first day of the
year with respect to the Julian calendar and added a sixth epagomenal day
every four years.

By default the invariant calendar is used, but the Sarkawag calendar can be
used beginning with the Armenian year 533 (11 August 1084) by passing the
parameter `method='sarkawag'` to the relevant functions.


### French Republican

Leap year calculations in the French Republican calendar are a matter of
dispute. By default, `convertdate` calculates leap years using the
autumnal equinox. You can also use one of three more systematic methods
proposed over the years.

-   Romme, a co-creator of the calendar, proposed leap years in years
    divisible by four, except for years divisible by 100.
-   Some concordances were drawn up in the 19th century that gave leap
    years every 4 years, in years that give a remainder of three when
    divided by four (19, 23, 27, etc...).
-   Von M&auml;dler proposed leap years in years divisible by four, except
    for years divisible by 128.

You can specify any of these three methods with the method keyword
argument in `french_republican` conversion functions.

    from convertdate import french_republican

    # Romme's method
    french_republican.to_gregorian(20, 1, 1), method='romme')
    # (1811, 9, 23)

    # continuous method
    french_republican.to_gregorian(20, 1, 1), method='continuous')
    # (1811, 9, 24)

    # von Madler's method
    french_republican.to_gregorian(20, 1, 1), method='madler')
    # (1811, 9, 23)

All the conversion methods correctly assign the leap years implemented
while calendar was in use (3, 7, 11).

Baha'i
------

The Bah&aacute;'&iacute; (Bad&iacute;) calendar has an intercalary period, Ayyam-i-H&aacute;, which occurs between the 18th and 19th months.
Dates in this period are returned as month 19, and the month of &lsquo;Al&aacute; is reported as month 20.

```python
from convertdate import bahai
# the first day of Ayyam-i-Ha:
bahai.to_gregorian(175, 19, 1)
# (2019, 2, 26)
# The first day of 'Ala:
bahai.to_gregorian(175, 20, 1)
# (2019, 3, 2)
```

Before the Common Era
---------------------

For dates before the Common Era (year 1), `convertdate` uses
astronomical notation: 1 BC is recorded as 0, 2 BC is -1, etc. This
makes arithmatic much easier at the expense of ignoring custom.

Note that for dates before 4 CE, `convertdate` uses the [proleptic
Julian
calendar](https://en.wikipedia.org/wiki/Proleptic_Julian_calendar). The
Julian Calendar was in use from 45 BC, but before 4 CE the leap year
leap year pattern was irregular.

The [proleptic Gregorian
calendar](https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar) is
used for dates before 1582 CE, the year of the Gregorian calendar
reform.

Holidays
--------

North American holidays are the current focus of the `holidays` module,
but pull requests are welcome.

    from convertdate import holidays

    # For simplicity, functions in the holidays module return a tuple
    # In the format (year, month, day)

    holidays.new_years(2014)
    # (2014, 1, 1)

    holidays.memorial_day(2014)
    # (2014, 5, 26)

    # USA is default
    holidays.thanksgiving(2014)
    # (2014, 11, 27)

    # But there is a Canadian option for some holidays
    holidays.thanksgiving(2014, 'canada')
    # (2014, 10, 13)

    # Mexican national holidays
    holidays.natalicio_benito_juarez(2016)
    # (2016, 3, 21)

    holidays.dia_revolucion(2016)
    # (2016, 11, 21)

    # Some Jewish holidays are included
    holidays.rosh_hashanah(2014)
    
    # Easter can be calculated according to different churches 
    # ('western', 'orthodox', 'eastern')
    # The eastern Christian computation differs from the Orthodox one
    # 4 times in each 532-year cycle.
    
    holidays.easter(2019)
    # (2019, 4, 21)
    holidays.easter(2019, church="orthodox")
    # (2019, 4, 28)
    holidays.easter(2019, church="orthodox")
    # (2019, 4, 28)

Utils
-----

Convertdate includes some utilities for manipulating and calculating
dates.

    from convertdate import utils

    # Calculate an arbitrary day of the week
    THUR = 3
    APRIL = 4

    # 3rd Thursday in April
    utils.nth_day_of_month(3, THUR, APRIL, 2014)
    # (2014, 4, 17)

    utils.nth_day_of_month(5, THUR, APRIL, 2014)
    # IndexError: No 5th day of month 4

    # Use 0 for the first argument to get the last weekday of a month
    utils.nth_day_of_month(0, THUR, APRIL, 2014)
    # (2014, 4, 24)

Note that when calculating weekdays, convertdate uses the convention of
the calendar and time modules: Monday is 0, Sunday is 6.

    from convertdate import gregorian

    SUN = 6

    day = gregorian.to_jd(2014, 4, 17)
    nextsunday = utils.next_weekday(SUN, day)

    gregorian.from_jd(nextsunday)
    # (2014, 4, 20)

Other utility functions:

-   nearest\_weekday
-   next\_or\_current\_weekday
-   previous\_weekday
-   previous\_or\_current\_weekday



HISTORY.rst
History
-------

2.3.0
------
* Fix bug in Bahai calculations during Ayyám-i-Há (#32). Thanks @chiuczek
* Regularize variable names:
 - coptic: rename `MONTH_NAMES` to `MONTHS`, rename `DAY_NAMES` to `WEEKDAYS`
 - french_republican: add `MONTHS`
 - mayan: rename `HAAB_MONTHS` to `HAAB`, rename `TZOLKIN_NAMES` to `TZOLKIN`
* Add some Jewish and Islamic holidays to `holidays`: `shemini_azeret`, `lag_baomer`, `tu_beshvat`, `tisha_bav`, `ramadan`, `ashura`, `eid_alfitr`, `eid_aladha`.
* Add docs. See https://convertdate.readthedocs.io/
* Fix December 31st bug in `ordinal` (#34).
* Add Indigenous Peoples' Day and Juneteenth to `holidays`. Deprecate Columbus day.

2.2.2
-----
* Add `observed` argument to the functions for several American holidays (#30)

2.2.1
-----
* Add orthodox/eastern Easter calculations and docs
* Add module for Armenian and Sarkawag regularisation
* Bump pytz requirement

2.2.0
-----
* Repair Bahai intercalary days bug (#13, thanks @bchurchill)
* Replace pyephem, which is now in maintenance mode, with pymeeus.
* Remove shebangs and regularize licenses (thanks @fabaff)
* Convert readme to ascii (#16)

2.1.3
-----
* Bump requirements

2.1.2
-----
* Additional Jewish holidays (thanks, @ohadeytan)
* Upload source distributions to Pypi (#10)

2.1.1
-----
* Add Coptic (Alexandrian) calendar converter.
* Add explicit support for Python 3.6.

2.1.0
-----
* Change Exception thrown on illegal dates to ValueError.
* Add Comte's Positivist calendar.
* Bump requirement versions.

2.0.8
-----
* Fix Persian weekday order (thanks, @meyt)

2.0.7
-----
* Better Python 2/3 compatibility
* Improve tests
* bump pytz requirement

2.0.6
-----
* Executing holidays module returns class with current year, not '2014'.
* Expand tests for French Republican, Bahai, Persian, holidays
* Add Travis CI testing

Bug fixes:

* Fix edge case when detecting the day of fall equinox in French Republican calendar
* Fix minor methods of calculating French Republican leap years.
* Change holidays.holidays.fathers_day from a method to a property
* Add pulaski_day to holidays.holidays
* Add pytz is dependency list

2.0.5
-----
* Fix Yom Kippur error in holidays.py (issue #3)

2.0.4
-----
* Typo in name of holidays.independence_day method
* Fix major bug in ordinal.from_gregorian
* Expand and organized tests

2.0.3.1
-------
Features:

* Add `ordinal` module, for counting the day of year
* Added Mexican national holidays
* Add `monthcalendar` functions

Other changes:

* Simplified logic in `ISO` module

2.0.3
-----
Features:

- Add list of day names and `day_name` function to French Republican converter
- Add multiple conversion methods to the French Republican calendar
- Add Dublin day count and Julian day count converters
- Add month names to Bahai and Hebrew calendars.

Other changes:

- Clarify that weekdays run Monday=0 to Sunday=6 (#2)
- Change Julian converter to use astronomical notation (0 = 1 BCE, -1 = 1 BCE)
- Expanded tests

2.0.2
-----
Features:

* Add support for Python 3 (#1)

tests/test_holidays.py
# -*- coding: utf-8 -*-
import unittest
from datetime import datetime
from convertdate import holidays, julian


class TestHolidays(unittest.TestCase):

    def setUp(self):
        self.h = holidays.Holidays(2015)

    def test_nth_day_of_month(self):
        assert holidays.nth_day_of_month(4, 2, 4, 2014) == (2014, 4, 23)
        self.assertRaises(IndexError, holidays.nth_day_of_month, 5, 3, 4, 2014)
        self.assertRaises(IndexError, holidays.nth_day_of_month, 6, 2, 4, 2014)
        self.assertRaises(IndexError, holidays.nth_day_of_month, 1, 7, 4, 2014)
        assert holidays.nth_day_of_month(4, 3, 11, 2014) == (2014, 11, 27)
        assert holidays.nth_day_of_month(0, 3, 11, 2014) == (2014, 11, 27)

    def test_holidays(self):
        h = holidays.Holidays(2014)
        self.assertEqual(h.christmas, (2014, 12, 25))
        assert h.thanksgiving == (2014, 11, 27)
        assert h.columbus_day == (2014, 10, 13)

        assert h.independence_day == (2014, 7, 4)

        assert self.h.christmas == (2015, 12, 25)
        assert self.h.christmas_eve == (2015, 12, 24)
        assert self.h.new_years == (2015, 1, 1)
        assert self.h.new_years_eve == (2015, 12, 31)
        assert self.h.valentines_day == (2015, 2, 14)
        assert self.h.halloween == (2015, 10, 31)
        assert self.h.mothers_day == (2015, 5, 10)
        self.assertEqual(self.h.fathers_day, (2015, 6, 21))

    def test_class(self):

        H = holidays.Holidays()
        assert H.year == datetime.now().year
        assert str(self.h) == 'Holidays(2015)'

        H.set_year(2010)

        assert H.year == 2010

    def test_events(self):
        assert holidays.new_years(2013) == (2013, 1, 1)
        assert holidays.martin_luther_king_day(2015) == (2015, 1, 19)

        assert holidays.lincolns_birthday(2015) == (2015, 2, 12)
        assert holidays.valentines_day(2015) == (2015, 2, 14)
        assert holidays.washingtons_birthday(2015) == (2015, 2, 22)
        assert holidays.presidents_day(2015) == (2015, 2, 16)

        assert holidays.pulaski_day(2015) == (2015, 3, 2)
        assert self.h.pulaski_day == (2015, 3, 2)

        assert holidays.may_day(2015) == (2015, 5, 1)

        assert holidays.columbus_day(2015, 'canada') == (2015, 10, 12)

        assert holidays.independence_day(2015) == (2015, 7, 4)
        assert holidays.independence_day(2015, True) == (2015, 7, 3)

    def test_thanksgiving(self):
        assert holidays.thanksgiving(2013) == (2013, 11, 28)
        assert holidays.thanksgiving(1939) == (1939, 11, 23)
        self.assertEqual(holidays.thanksgiving(1941), (1941, 11, 20))

        assert self.h.thanksgiving == (2015, 11, 26)

        assert holidays.thanksgiving(2015, 'canada') == (2015, 10, 12)

    def test_easterWestern(self):
        easters = [
            (1994, 4, 3),
            (1995, 4, 16),
            (1996, 4, 7),
            (1997, 3, 30),
            (1998, 4, 12),
            (1999, 4, 4),
            (2000, 4, 23),
            (2001, 4, 15),
            (2002, 3, 31),
            (2003, 4, 20),
            (2004, 4, 11),
            (2005, 3, 27),
            (2006, 4, 16),
            (2007, 4, 8),
            (2008, 3, 23),
            (2009, 4, 12),
            (2010, 4, 4),
            (2011, 4, 24),
            (2012, 4, 8),
            (2013, 3, 31),
            (2014, 4, 20),
            (2015, 4, 5),
            (2016, 3, 27),
            (2017, 4, 16),
            (2018, 4, 1),
            (2019, 4, 21),
            (2020, 4, 12),
            (2021, 4, 4),
            (2022, 4, 17),
            (2023, 4, 9),
            (2024, 3, 31),
            (2025, 4, 20),
            (2026, 4, 5),
            (2027, 3, 28),
            (2028, 4, 16),
            (2029, 4, 1),
            (2030, 4, 21),
            (2031, 4, 13),
            (2032, 3, 28),
            (2033, 4, 17),
            (2034, 4, 9),
            (2345, 4, 22)
        ]

        for y, m, d in easters:
            self.assertEqual(holidays.easter(y), (y, m, d))

    def test_easterEastern(self):
        easters = [
            (1999, 4, 11),
            (2000, 4, 30),
            (2001, 4, 15),
            (2002, 5, 5),
            (2003, 4, 27),
            (2004, 4, 11),
            (2005, 5, 1),
            (2006, 4, 23),
            (2007, 4, 8),
            (2008, 4, 27),
            (2009, 4, 19),
            (2010, 4, 4),
            (2011, 4, 24),
            (2012, 4, 15),
            (2013, 5, 5),
            (2014, 4, 20),
            (2015, 4, 12),
            (2016, 5, 1),
            (2017, 4, 16),
            (2018, 4, 8),
            (2019, 4, 28),
            (2020, 4, 19),
            (2021, 5, 2),
            (2022, 4, 24),
            (2023, 4, 16),
            (2024, 5, 5),
            (2025, 4, 20),
            (2026, 4, 12),
            (2027, 5, 2),
            (2028, 4, 16),
            (2029, 4, 8),
            (2030, 4, 28),
            (2031, 4, 13),
            (2032, 5, 2),
            (2033, 4, 24),
            (2034, 4, 9),
            (2035, 4, 29),
            (2036, 4, 20),
            (2037, 4, 5),
            (2038, 4, 25),
            (2039, 4, 17),
            (2056, 4, 9),
            (2156, 4, 11)
        ]

        for y, m, d in easters:
            self.assertEqual(holidays.easter(y, "orthodox"), (y, m, d))
            self.assertEqual(holidays.easter(y, "eastern"), (y, m, d))

        self.assertEqual(self.h.easter, (2015, 4, 5))

    def testNonChalcedonian(self):
        # In these years, Orthodox Easter falls on 6 April (Julian),
        # but Non-Chalcedonian churches celebrate it a week later on 13 Aprail
        years = 570, 665, 760, 1007, 1102, 1197, 1292, 1539, 1634, 1729, 1824, 2071, 2166, 2261, 2356

        for y in years:
            orthodox = julian.from_gregorian(*holidays.easter(y, "orthodox"))
            eastern = julian.from_gregorian(*holidays.easter(y, "eastern"))
            self.assertNotEqual(orthodox, eastern)
            self.assertEqual((y, 4, 6), orthodox)
            self.assertEqual((y, 4, 13), eastern)

        for y in years:
            self.assertEqual(holidays.easter(y + 1, "orthodox"), holidays.easter(y + 1, "eastern"))


    def test_jewish_holidays(self):
        # http://www.chabad.org/holidays/passover/pesach_cdo/aid/671901/jewish/When-is-Passover-in-2013-2014-2015-2016-and-2017.htm
        # the date here is the start of the holiday, so the eve=1 option is used
        passovers = [
            (2013, 3, 25),
            (2014, 4, 14),
            (2015, 4, 3),
            (2016, 4, 22),
            (2017, 4, 10)
        ]
        for y, m, d in passovers:
            self.assertEqual(holidays.passover(y, eve=1), (y, m, d))

        rosh_hashanahs = [
            (2014, 9, 24),
            (2015, 9, 13),
            (2016, 10, 2),
            (2017, 9, 20),
        ]
        for y, m, d in rosh_hashanahs:
            self.assertEqual(holidays.rosh_hashanah(y, eve=1), (y, m, d))

        self.assertEqual(holidays.hanukkah(2015, True), (2015, 12, 6))
        self.assertEqual(holidays.hanukkah(2015), (2015, 12, 7))

        self.assertEqual(holidays.yom_kippur(2015), (2015, 9, 23))
        self.assertEqual(holidays.yom_kippur(2015, True), (2015, 9, 22))

        sukkots = [
            (2016, 10, 17),
            (2015, 9, 28),
        ]
        for y, m, d in sukkots:
            self.assertEqual(holidays.sukkot(y, eve=0), (y, m, d))

        shavuots = [
            (2016, 6, 12),
            (2015, 5, 24)
        ]
        for y, m, d in shavuots:
            self.assertEqual(holidays.shavuot(y, eve=0), (y, m, d))

        purims = [
            (2017, 3, 12),
            (2016, 3, 24)
        ]
        for y, m, d in purims:
            self.assertEqual(holidays.purim(y, eve=0), (y, m, d))

        tisha_bavs = [
            (2019, 8, 11),
            (2020, 7, 30),
            (2021, 7, 18),
            (2022, 8, 7),
            (2023, 7, 27),
        ]
        for y, m, d in tisha_bavs:
            self.assertEqual(holidays.tisha_bav(y, eve=0), (y, m, d))

        assert self.h.hanukkah == (2015, 12, 7)
        assert self.h.rosh_hashanah == (2015, 9, 14)
        assert self.h.yom_kippur == (2015, 9, 23)
        assert self.h.passover == (2015, 4, 4)

        assert self.h.tisha_bav == (2015, 7, 26)
        assert self.h.shemini_azeret == (2015, 10, 5)
        assert self.h.lag_baomer == (2015, 5, 7)
        assert self.h.tu_beshvat == (2015, 2, 4)

    def test_mexican_holidays(self):
        self.assertEqual(holidays.natalicio_benito_juarez(2015, False), (2015, 3, 21))
        self.assertEqual(holidays.natalicio_benito_juarez(2015), (2015, 3, 16))

        assert self.h.dia_constitucion == (2015, 2, 2)
        assert self.h.natalicio_benito_juarez == (2015, 3, 16)
        assert self.h.dia_independencia == (2015, 9, 16)
        assert self.h.dia_revolucion == (2015, 11, 16)

    def test_usa_holidays(self):
        assert self.h.independence_day == (2015, 7, 3)
        assert self.h.flag_day == (2015, 6, 14)
        assert self.h.election_day == (2015, 11, 3)
        assert self.h.presidents_day == (2015, 2, 16)
        assert self.h.washingtons_birthday == (2015, 2, 22)
        assert self.h.lincolns_birthday == (2015, 2, 12)
        assert self.h.memorial_day == (2015, 5, 25)
        assert self.h.labor_day == (2015, 9, 7)
        assert self.h.columbus_day == (2015, 10, 12)
        assert self.h.veterans_day == (2015, 11, 11)
        assert self.h.martin_luther_king_day == (2015, 1, 19)


if __name__ == '__main__':
    unittest.main()


src/convertdate/holidays.py
# -*- coding: utf-8 -*-
# This file is part of convertdate.
# http://github.com/fitnr/convertdate
# Licensed under the MIT license:
# http://opensource.org/licenses/MIT
# Copyright (c) 2016, fitnr <fitnr@fakeisthenewreal>
"""
Generate the dates of common North American and Jewish holidays.

Jewish holidays begin the sunset before the first (secular) day of the holiday
With the ``eve`` option set to ``True``, the day of this sunset is returned
without the ``eve`` option, the (secular) day is returned. This is the default.
"""
import calendar
import time
from math import trunc
from .utils import jwday, nth_day_of_month
from . import gregorian, hebrew, islamic, julian

# weekdays
MON = 0
TUE = 1
WED = 2
THU = 3
FRI = 4
SAT = 5
SUN = 6

# months
JAN = 1
FEB = 2
MAR = 3
APR = 4
MAY = 5
JUN = 6
JUL = 7
AUG = 8
SEP = 9
OCT = 10
NOV = 11
DEC = 12


def new_years(year, observed=None):
    '''Jan 1st, possibly observed on last day of previous year'''
    if observed:
        weekday = calendar.weekday(year, JAN, 1)
        if weekday == SAT:
            return (year - 1, DEC, 31)
        if weekday == SUN:
            return (year, JAN, 2)
    return (year, JAN, 1)


def martin_luther_king_day(year):
    '''third monday in January'''
    return nth_day_of_month(3, MON, JAN, year)


def lincolns_birthday(year):
    '''Feb 12'''
    return (year, FEB, 12)


def valentines_day(year):
    '''feb 14th'''
    return (year, FEB, 14)


def washingtons_birthday(year, observed=None):
    '''Feb 22, possibly observed on 3rd Monday in February'''
    if observed:
        return nth_day_of_month(3, MON, FEB, year)
    return (year, FEB, 22)


def presidents_day(year):
    '''3rd Monday of Feb'''
    return nth_day_of_month(3, MON, FEB, year)


def pulaski_day(year):
    '''1st monday in March'''
    return nth_day_of_month(1, MON, MAR, year)


def easter(year, church=None):
    '''Calculate Easter in the given church according to the given calendar.'''
    church = church or "western"

    if church == "western":
        return _easter_western(year)

    if church == "orthodox":
        return julian.to_gregorian(*_easter_julian(year))

    if church == "eastern":
        return julian.to_gregorian(*_easter_julian(year, mode="eastern"))

    raise ValueError("Unknown value for 'church'")


def _easter_western(year):
    '''Calculate western easter'''
    # formula taken from http://aa.usno.navy.mil/faq/docs/easter.html
    c = trunc(year / 100)
    n = year - 19 * trunc(year / 19)

    k = trunc((c - 17) / 25)

    i = c - trunc(c / 4) - trunc((c - k) / 3) + (19 * n) + 15
    i = i - 30 * trunc(i / 30)
    i = i - trunc(i / 28) * (1 - trunc(i / 28) * trunc(29 / (i + 1)) * trunc((21 - n) / 11))

    j = year + trunc(year / 4) + i + 2 - c + trunc(c / 4)
    j = j - 7 * trunc(j / 7)

    L = i - j

    month = 3 + trunc((L + 40) / 44)
    day = L + 28 - 31 * trunc(month / 4)

    date = (year, int(month), int(day))

    return date


def _easter_julian(year, mode="dionysian"):
    '''Calculate Easter for the orthodox and eastern churches in the Julian calendar.'''
    # Uses Meeus's Julian algorithm.
    meton = year % 19
    b = year % 4
    c = year % 7
    d = (19 * meton + 15) % 30
    if mode == "eastern" and meton == 0:
        d = d + 1
    e = (2 * b + 4 * c - d + 6) % 7
    fmj = 113 + d  # Easter full moon (days after -92 March)
    dmj = fmj + e + 1  # Easter Sunday (days after -92 March)
    esmj = trunc(dmj / 31)  # month of Easter Sunday
    esdj = (dmj % 31) + 1  # day of Easter Sunday

    return year, esmj, esdj


def may_day(year):
    return (year, MAY, 1)


def mothers_day(year):
    '''2nd Sunday in May'''
    return nth_day_of_month(2, SUN, MAY, year)


def memorial_day(year):
    '''last Monday in May'''
    return nth_day_of_month(0, MON, MAY, year)


def fathers_day(year):
    '''3rd Sunday in June'''
    return nth_day_of_month(3, SUN, JUN, year)


def juneteenth(year):
    '''19th of June'''
    return year, JUN, 19

def flag_day(year):
    '''June 14th'''
    return (year, JUN, 14)


def independence_day(year, observed=None):
    '''July 4th'''
    day = 4

    if observed:
        weekday = calendar.weekday(year, JUL, 4)
        if weekday == SAT:
            day = 3
        if weekday == SUN:
            day = 5

    return (year, JUL, day)


def labor_day(year):
    '''first Monday in Sep'''
    return nth_day_of_month(1, MON, SEP, year)


def indigenous_peoples_day(year, country='usa'):
    """in USA: 2nd Monday in Oct
    Elsewhere: Oct 12"""
    if country == 'usa':
        return nth_day_of_month(2, MON, OCT, year)

    return (year, OCT, 12)


def columbus_day(year, country='usa'):
    raise DeprecationWarning("columbus_day will be removed in a future release")
    return indigenous_peoples_day(year, country)


def halloween(year):
    '''Oct 31'''
    return (year, OCT, 31)


def election_day(year):
    '''1st Tues in Nov'''
    return nth_day_of_month(1, TUE, NOV, year)


def veterans_day(year, observed=None):
    '''Nov 11, or the following closest weekday'''
    day = 11
    if observed:
        weekday = calendar.weekday(year, NOV, 11)
        if weekday == SAT:
            day = 10
        if weekday == SUN:
            day = 12

    return (year, NOV, day)


def rememberance_day(year):
    return veterans_day(year)


def armistice_day(year):
    return veterans_day(year)


def thanksgiving(year, country='usa'):
    '''USA: last Thurs. of November, Canada: 2nd Mon. of October'''
    if country == 'usa':
        if year in [1940, 1941]:
            return nth_day_of_month(3, THU, NOV, year)
        elif year == 1939:
            return nth_day_of_month(4, THU, NOV, year)

        return nth_day_of_month(0, THU, NOV, year)

    if country == 'canada':
        return nth_day_of_month(2, MON, OCT, year)

    raise NotImplementedError('Unsupported argument for country')


def christmas_eve(year):
    '''24th of December'''
    return (year, DEC, 24)


def christmas(year, observed=None):
    '''25th of December'''
    day = 25
    if observed:
        weekday = calendar.weekday(year, DEC, 25)
        if weekday == SAT:
            day = 24
        if weekday == SUN:
            day = 26
    return (year, DEC, day)


def new_years_eve(year):
    '''Dec 31st'''
    return (year, DEC, 31)


def hanukkah(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.KISLEV, 25)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def purim(year, eve=None):
    if not hebrew.leap(year + hebrew.HEBREW_YEAR_OFFSET):
        jd = hebrew.to_jd_gregorianyear(year, hebrew.ADAR, 14)
    else:
        jd = hebrew.to_jd_gregorianyear(year, hebrew.VEADAR, 14)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def rosh_hashanah(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.TISHRI, 1)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def yom_kippur(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.TISHRI, 10)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def passover(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.NISAN, 15)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def shavuot(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.SIVAN, 6)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def sukkot(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.TISHRI, 15)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def shemini_azeret(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.TISHRI, 22)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def lag_baomer(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.IYYAR, 18)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def tu_beshvat(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.SHEVAT, 15)
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


def tisha_bav(year, eve=None):
    jd = hebrew.to_jd_gregorianyear(year, hebrew.AV, 9)
    if jwday(jd) == SAT:
        jd = jd + 1
    if eve:
        jd = jd - 1
    return gregorian.from_jd(jd)


# Mexican holidays


def dia_constitucion(year, observed=True):
    """
    Constitution Day, a public holiday in Mexico observed
    on the first Monday in February.
    """
    if observed:
        return nth_day_of_month(1, MON, FEB, year)

    return (year, FEB, 5)


def natalicio_benito_juarez(year, observed=True):
    """
    Benito Juárez's Birthday, a public holiday in Mexico observed
    on the third Monday in March.
    """
    if observed:
        return nth_day_of_month(3, MON, MAR, year)

    return (year, MAR, 21)


def dia_independencia(year):
    """
    Mexican independence day, observed on September 16.
    """
    return year, SEP, 16


def dia_revolucion(year):
    """
    Revolution Day, a public holiday in Mexico observed
    on the third Monday in November.
    """
    return nth_day_of_month(3, MON, NOV, year)


# Islamic holidays

def ramadan(year, eve=None):
    jd = islamic.to_jd_gregorianyear(year, 9, 1)
    if eve:
        jd = jd = 1
    return gregorian.from_jd(jd)


def ashura(year, eve=None):
    jd = islamic.to_jd_gregorianyear(year, 1, 10)
    if eve:
        jd = jd = 1
    return gregorian.from_jd(jd)


def eid_alfitr(year, eve=None):
    jd = islamic.to_jd_gregorianyear(year, 10, 1)
    if eve:
        jd = jd = 1
    return gregorian.from_jd(jd)


def eid_aladha(year, eve=None):
    jd = islamic.to_jd_gregorianyear(year, 12, 10)
    if eve:
        jd = jd = 1
    return gregorian.from_jd(jd)


class Holidays:
    '''Convenience class for fetching many holidays in a given year.'''
    # pylint: disable=missing-function-docstring

    def __init__(self, year=None):
        self.year = year or time.localtime().tm_year

    def set_year(self, year):
        self.year = year

    def __repr__(self):
        return 'Holidays({})'.format(self.year)

    # the holidays...
    @property
    def christmas(self):
        return christmas(self.year, True)

    @property
    def christmas_eve(self):
        return christmas_eve(self.year)

    @property
    def thanksgiving(self):
        return thanksgiving(self.year)

    @property
    def new_years(self):
        return new_years(self.year, True)

    @property
    def new_years_eve(self):
        return new_years_eve(self.year)

    @property
    def independence_day(self):
        return independence_day(self.year, observed=True)

    @property
    def flag_day(self):
        return flag_day(self.year)

    @property
    def election_day(self):
        return election_day(self.year)

    @property
    def presidents_day(self):
        return presidents_day(self.year)

    @property
    def washingtons_birthday(self):
        return washingtons_birthday(self.year)

    @property
    def lincolns_birthday(self):
        return lincolns_birthday(self.year)

    @property
    def memorial_day(self):
        return memorial_day(self.year)

    @property
    def juneteenth(self):
        return juneteenth(self.year)

    @property
    def labor_day(self):
        return labor_day(self.year)

    @property
    def indigenous_peoples_day(self):
        return indigenous_peoples_day(self.year)

    @property
    def columbus_day(self):
        return indigenous_peoples_day(self.year)

    @property
    def veterans_day(self):
        return veterans_day(self.year, True)

    @property
    def valentines_day(self):
        return valentines_day(self.year)

    @property
    def halloween(self):
        return halloween(self.year)

    @property
    def mothers_day(self):
        return mothers_day(self.year)

    @property
    def fathers_day(self):
        return fathers_day(self.year)

    @property
    def pulaski_day(self):
        return pulaski_day(self.year)

    @property
    def easter(self):
        return easter(self.year)

    @property
    def martin_luther_king_day(self):
        return martin_luther_king_day(self.year)

    @property
    def hanukkah(self):
        return hanukkah(self.year, eve=False)

    @property
    def purim(self):
        return purim(self.year, eve=False)

    @property
    def rosh_hashanah(self):
        return rosh_hashanah(self.year, eve=False)

    @property
    def yom_kippur(self):
        return yom_kippur(self.year, eve=False)

    @property
    def passover(self):
        return passover(self.year, eve=False)

    @property
    def shavuot(self):
        return shavuot(self.year, eve=False)

    @property
    def sukkot(self):
        return sukkot(self.year, eve=False)

    @property
    def tu_beshvat(self):
        return tu_beshvat(self.year, eve=False)

    @property
    def shemini_azeret(self):
        return shemini_azeret(self.year, eve=False)

    @property
    def lag_baomer(self):
        return lag_baomer(self.year, eve=False)

    @property
    def tisha_bav(self):
        return tisha_bav(self.year, eve=False)

    @property
    def dia_constitucion(self):
        return dia_constitucion(self.year, observed=True)

    @property
    def natalicio_benito_juarez(self):
        return natalicio_benito_juarez(self.year, observed=True)

    @property
    def dia_independencia(self):
        return dia_independencia(self.year)

    @property
    def dia_revolucion(self):
        return dia_revolucion(self.year)

    @property
    def ramadan(self):
        return ramadan(self.year)

    @property
    def ashura(self):
        return ashura(self.year)

    @property
    def eid_alfitr(self):
        return eid_alfitr(self.year)

    @property
    def eid_aladha(self):
        return eid_aladha(self.year)


if __name__ == '__main__':
    holiday = Holidays(time.localtime().tm_year)
